File: examples/grammy-bot/src/bot.ts
/**
 * A small, realistic grammY bot with FlowCastle attached.
 *
 * Everything below /start is ordinary grammY code you already know how to write.
 * FlowCastle is installed once with `bot.use(fc)` and then used from handlers via
 * `ctx.flowcastle` — identify a contact, record goals, hand a chat to a human, or
 * start a flow that teammates built visually in the FlowCastle editor.
 */
import { Bot, Context, InlineKeyboard } from 'grammy';
import { flowcastle, FlowCastleFlavor } from '@flowcastle/grammy';

import {
  completedQualification,
  InMemoryLeadQualificationStore,
  nextQualificationStep,
  parseQualificationCallback,
  QUALIFICATION_OPTIONS,
  QUALIFICATION_QUESTIONS,
  qualificationLabel,
  saveQualificationAnswer,
  type QualificationStep,
} from './lead-qualification';

type BotContext = FlowCastleFlavor<Context>;

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value) {
    console.error(`Missing required environment variable ${name}.`);
    process.exit(1);
  }
  return value;
}

const BOT_TOKEN = requireEnv('BOT_TOKEN');
const FLOWCASTLE_API_KEY = requireEnv('FLOWCASTLE_API_KEY');
const FOLLOW_UP_FLOW_KEY = process.env.FLOWCASTLE_FOLLOW_UP_FLOW_KEY;

const bot = new Bot<BotContext>(BOT_TOKEN);

// Install FlowCastle FIRST so it observes every update and outgoing call.
const fc = flowcastle<BotContext>({
  apiKey: FLOWCASTLE_API_KEY,
  // Routing-only content: commands and callback ids are shared, free text is not.
  // Switch to `{ messageContent: 'full' }` if your FlowCastle flows need message text.
  privacy: { contactFields: ['username', 'languageCode'] },
  // Lets flows built in the FlowCastle editor run through this bot process.
  runtime: { enabled: true },
  onError: (error) => console.error('[flowcastle]', error),
});
bot.use(fc);

const qualification = new InMemoryLeadQualificationStore();

const keyboardFor = (step: QualificationStep): InlineKeyboard => {
  const keyboard = new InlineKeyboard();
  for (const option of QUALIFICATION_OPTIONS[step]) keyboard.text(option.label, `lead:${step}:${option.value}`).row();
  return keyboard;
};

bot.command('start', async (ctx) => {
  const displayName = [ctx.from?.first_name, ctx.from?.last_name].filter(Boolean).join(' ') || ctx.from?.username;
  ctx.flowcastle.identify({ displayName });
  await ctx.reply(
    `Welcome${ctx.from?.first_name ? `, ${ctx.from.first_name}` : ''}! This is the FlowCastle grammY example.\n\n` +
      'Send any text and I will echo it back, or tap the button to record a demo goal.',
    { reply_markup: new InlineKeyboard().text('🎯 Record demo goal', 'demo_goal') },
  );
});

bot.command('help', (ctx) =>
  ctx.reply(
    '/start — greet you and identify you as a FlowCastle contact\n' +
      '/qualify — three-step lead qualification, fully in code\n' +
      '/human — hand this chat to a person in FlowCastle Live Chat\n' +
      '/help — this message',
  ),
);

bot.command('qualify', async (ctx) => {
  if (ctx.from === undefined) return;
  qualification.set(ctx.from.id, {});
  await ctx.reply(QUALIFICATION_QUESTIONS.need, { reply_markup: keyboardFor('need') });
});

bot.command('human', async (ctx) => {
  ctx.flowcastle.requestLiveAgent({ note: 'User asked to talk to a human.' });
  await ctx.reply('Connecting you with a teammate — they will reply right here. 💬');
});

bot.on('callback_query:data', async (ctx) => {
  if (ctx.callbackQuery.data === 'demo_goal') {
    ctx.flowcastle.goal('demo_goal', { source: 'example' });
    await ctx.answerCallbackQuery({ text: 'Goal recorded!' });
    await ctx.reply('Recorded a demo_goal — check Analytics in your FlowCastle dashboard.');
    return;
  }

  const answer = parseQualificationCallback(ctx.callbackQuery.data);
  if (answer === undefined) return ctx.answerCallbackQuery();

  const current = qualification.get(ctx.from.id);
  const updated = current === undefined ? undefined : saveQualificationAnswer(current, answer.step, answer.value);
  if (updated === undefined) return ctx.answerCallbackQuery({ text: 'This step expired. Send /qualify to restart.' });

  qualification.set(ctx.from.id, updated);
  await ctx.answerCallbackQuery({ text: 'Saved' });

  const nextStep = nextQualificationStep(updated);
  if (nextStep !== undefined) return ctx.reply(QUALIFICATION_QUESTIONS[nextStep], { reply_markup: keyboardFor(nextStep) });

  const result = completedQualification(updated);
  if (result === undefined) return;
  const lead = { leadNeed: result.need, leadBudget: result.budget, leadTimeline: result.timeline };
  qualification.delete(ctx.from.id);

  // Answers become contact traits (visible in the CRM) and a goal (visible in funnels).
  ctx.flowcastle.identify(lead);
  ctx.flowcastle.goal('lead_qualified', lead);
  await ctx.reply(
    'Thanks — your answers are saved.\n\n' +
      `Need: ${qualificationLabel('need', lead.leadNeed)}\n` +
      `Budget: ${qualificationLabel('budget', lead.leadBudget)}\n` +
      `Timeline: ${qualificationLabel('timeline', lead.leadTimeline)}`,
  );

  // Optional: hand off to a follow-up flow that a teammate built in the FlowCastle editor.
  if (FOLLOW_UP_FLOW_KEY !== undefined) {
    try {
      await ctx.flowcastle.runFlow(FOLLOW_UP_FLOW_KEY, { inputs: lead });
    } catch (error) {
      console.error('[flowcastle] follow-up flow failed to start', error);
    }
  }
});

bot.on('message:text', async (ctx) => {
  // A human is handling this chat in Live Chat — do not echo over them.
  if (ctx.flowcastle.isLiveAgentActive) return;
  await ctx.reply(`Echo: ${ctx.message.text}`);
});

async function shutdown(signal: NodeJS.Signals): Promise<void> {
  console.log(`\n${signal}: shutting down`);
  await bot.stop();
  await fc.flush(); // ship anything still buffered
  fc.destroy();
  process.exit(0);
}
process.once('SIGINT', () => void shutdown('SIGINT'));
process.once('SIGTERM', () => void shutdown('SIGTERM'));

async function main(): Promise<void> {
  await fc.ready(); // load the flow manifest before polling so the first matching update is routed correctly
  await bot.start({ onStart: (me) => console.log(`@${me.username} is up with FlowCastle attached.`) });
}
void main().catch((error: unknown) => {
  console.error('Failed to start:', error);
  fc.destroy();
  process.exitCode = 1;
});

File: examples/aiogram-bot/bot.py
"""A small, realistic aiogram 3 bot with FlowCastle attached.

Everything below /start is ordinary aiogram code. FlowCastle is installed once with
`adapter.install(dp)` and then used from handlers through the `flowcastle` argument
(injected via middleware data) — identify a contact, record goals, hand a chat to a
human, or start a flow that teammates built visually in the FlowCastle editor.
"""

from __future__ import annotations

import asyncio
import logging
import os
import sys

from aiogram import Bot, Dispatcher, F
from aiogram.filters import Command, CommandStart
from aiogram.types import CallbackQuery, InlineKeyboardButton, InlineKeyboardMarkup, Message
from dotenv import load_dotenv
from flowcastle import FlowCastleContext, FlowCastleCore, FlowCastleOptions, PrivacyOptions
from flowcastle.adapters.aiogram import AiogramAdapter

from lead_qualification import OPTIONS, QUESTIONS, InMemoryStore, label, parse_callback

load_dotenv()
logging.basicConfig(level=logging.INFO)


def require_env(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        sys.exit(f'Missing required environment variable {name}.')
    return value


BOT_TOKEN = require_env('BOT_TOKEN')
FLOWCASTLE_API_KEY = require_env('FLOWCASTLE_API_KEY')
FOLLOW_UP_FLOW_KEY = os.environ.get('FLOWCASTLE_FOLLOW_UP_FLOW_KEY')

core = FlowCastleCore(FlowCastleOptions(
    api_key=FLOWCASTLE_API_KEY,
    # Routing-only content: commands and callback ids are shared, free text is not.
    # Use message_content='full' if your FlowCastle flows need message text.
    privacy=PrivacyOptions(contact_fields=('username', 'language_code')),
    # Lets flows built in the FlowCastle editor run through this bot process.
    runtime_enabled=True,
))
adapter = AiogramAdapter(core)

dp = Dispatcher()
qualification = InMemoryStore()


def keyboard_for(step: str) -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(inline_keyboard=[
        [InlineKeyboardButton(text=text, callback_data=f'lead:{step}:{value}')] for text, value in OPTIONS[step]
    ])


@dp.message(CommandStart())
async def start(message: Message, flowcastle: FlowCastleContext) -> None:
    user = message.from_user
    display_name = ' '.join(filter(None, [user.first_name, user.last_name])) if user else None
    await flowcastle.identify({'displayName': display_name})
    await message.answer(
        f'Welcome, {user.first_name if user else "there"}! This is the FlowCastle aiogram example.\n\n'
        'Send any text and I will echo it back, or tap the button to record a demo goal.',
        reply_markup=InlineKeyboardMarkup(inline_keyboard=[[InlineKeyboardButton(text='🎯 Record demo goal', callback_data='demo_goal')]]),
    )


@dp.message(Command('help'))
async def help_command(message: Message) -> None:
    await message.answer(
        '/start — greet you and identify you as a FlowCastle contact\n'
        '/qualify — three-step lead qualification, fully in code\n'
        '/human — hand this chat to a person in FlowCastle Live Chat\n'
        '/help — this message'
    )


@dp.message(Command('qualify'))
async def qualify(message: Message) -> None:
    if message.from_user is None:
        return
    qualification.start(message.from_user.id)
    await message.answer(QUESTIONS['need'], reply_markup=keyboard_for('need'))


@dp.message(Command('human'))
async def human(message: Message, flowcastle: FlowCastleContext) -> None:
    await flowcastle.request_live_agent('User asked to talk to a human.')
    await message.answer('Connecting you with a teammate — they will reply right here. 💬')


@dp.callback_query(F.data == 'demo_goal')
async def demo_goal(query: CallbackQuery, flowcastle: FlowCastleContext) -> None:
    await flowcastle.goal('demo_goal', {'source': 'example'})
    await query.answer('Goal recorded!')
    if isinstance(query.message, Message):
        await query.message.answer('Recorded a demo_goal — check Analytics in your FlowCastle dashboard.')


@dp.callback_query(F.data.startswith('lead:'))
async def qualification_answer(query: CallbackQuery, flowcastle: FlowCastleContext) -> None:
    parsed = parse_callback(query.data or '')
    state = qualification.get(query.from_user.id)
    if parsed is None or state is None or not state.save(*parsed):
        await query.answer('This step expired. Send /qualify to restart.')
        return
    await query.answer('Saved')
    if not isinstance(query.message, Message):
        return

    next_step = state.next_step()
    if next_step is not None:
        await query.message.answer(QUESTIONS[next_step], reply_markup=keyboard_for(next_step))
        return

    lead = state.completed()
    if lead is None:
        return
    qualification.delete(query.from_user.id)

    # Answers become contact traits (visible in the CRM) and a goal (visible in funnels).
    await flowcastle.identify(lead)
    await flowcastle.goal('lead_qualified', lead)
    await query.message.answer(
        'Thanks — your answers are saved.\n\n'
        f"Need: {label('need', lead['leadNeed'])}\n"
        f"Budget: {label('budget', lead['leadBudget'])}\n"
        f"Timeline: {label('timeline', lead['leadTimeline'])}"
    )

    # Optional: hand off to a follow-up flow that a teammate built in the FlowCastle editor.
    if FOLLOW_UP_FLOW_KEY:
        try:
            await flowcastle.run_flow(FOLLOW_UP_FLOW_KEY, lead)
        except Exception:  # noqa: BLE001 — never let FlowCastle break the bot
            logging.exception('[flowcastle] follow-up flow failed to start')


@dp.message(F.text)
async def echo(message: Message) -> None:
    await message.answer(f'Echo: {message.text}')


async def main() -> None:
    bot = Bot(BOT_TOKEN)
    await adapter.ready()  # load the flow manifest before polling
    adapter.install(dp)
    dp.shutdown.register(adapter.stop)  # final bounded flush on exit
    logging.info('aiogram bot is up with FlowCastle attached.')
    await dp.start_polling(bot)


if __name__ == '__main__':
    asyncio.run(main())
